packages/data-provider/src/utils/fileUtils.ts

```diff
@@ -33,43 +78,73 @@ export const readHeaders = async (file: Express.Multer.File) => {
 };
 
 /**
- * Reads a .tsv file and parse it to a JSON format
- * @param {string} fileName Full filename path of the .tsv file
+ * Reads a text file and parse it to a JSON format.
+ * Records are parsed to match schema field types.
+ * Supported files: .tsv and .csv
+ * @param {Express.Multer.File} file A file to read
  * @param {Schema} schema Schema to parse data with
  * @returns a JSON format objet
  */
-export const tsvToJson = async (
-	fileName: string,
+export const readTextFile = async (
+	file: Express.Multer.File,
 	schema: Schema,
 ): Promise<{ records: DataRecord[]; errors?: ParseSchemaError[] }> => {
-	const contents = await fsPromises.readFile(fileName, 'utf-8');
-	const arr = parseTsvToJson(contents);
-	const parseSchemaResult = parse.parseSchemaValues(arr, schema);
-	if (parseSchemaResult.success) {
-		return { records: parseSchemaResult.data.records };
+	const returnRecords: DataRecord[] = [];
+	const returnErrors: ParseSchemaError[] = [];
+	const separatorCharacter = getSeparatorCharacter(file.originalname);
+	if (!separatorCharacter) {
+		throw new Error('Invalid file Extension');
 	}
-	return {
-		records: parseSchemaResult.data.records,
-		errors: parseSchemaResult.data.errors,
-	};
-};
 
-const parseTsvToJson = (content: string): UnprocessedDataRecord[] => {
-	const lines = content.split('\n');
-	const headers = lines.slice(0, 1)[0].trim().split('\t');
-	const rows = lines
-		.slice(1, lines.length)
-		.filter((line) => line && line.trim() !== '')
-		.map((line) => {
-			const data = line.split('\t');
-			return headers.reduce((obj: UnprocessedDataRecord, nextKey, index) => {
-				const dataStr = data[index] || '';
-				const formattedData = formatForExcelCompatibility(dataStr);
-				obj[nextKey] = formattedData;
-				return obj;
-			}, {});
+	let headers: string[] = [];
+	let lineNumber = 0;
+
+	return new Promise((resolve, reject) => {
+		const stream = fs.createReadStream(file.path).pipe(csvParse({ delimiter: separatorCharacter }));
+
+		stream.on('data', (record: string[]) => {
+			lineNumber++;
+			if (!headers.length) {
+				headers = Object.values(record);
+			} else {
+				const mappedRecord = mapRecordToHeaders(headers, record);
+
+				try {
+					const parseSchemaResult = parse.parseRecordValues(mappedRecord, schema);
+					if (parseSchemaResult.success) {
+						returnRecords.push(parseSchemaResult.data.record);
+					} else {
+						returnRecords.push(parseSchemaResult.data.record);
+						returnErrors.push({
+							recordErrors: parseSchemaResult.data.errors,
+							recordIndex: lineNumber,
+						});
+					}
+
+					if (lineNumber % 1000 === 0) {
+						// TODO: Add batch processing logic here (e.g., write to database or process as needed)
+						// returnRecords = []; // Clear the array after processing the batch
+						// returnErrors = []; // Clear the array after processing the batch
+					}
```

Comment:
Adding a placeholder here. Integration with Maestro to index records in ES is implemented in a separate PR

Reply:
This is reading in data from a file, what maestro integration is required at this stage?

Reply:
you're right this is not meant to be indexed, just processed for schema validation.